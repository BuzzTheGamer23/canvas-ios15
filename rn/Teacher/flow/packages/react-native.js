// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
