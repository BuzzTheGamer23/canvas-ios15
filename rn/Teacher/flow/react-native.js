//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
