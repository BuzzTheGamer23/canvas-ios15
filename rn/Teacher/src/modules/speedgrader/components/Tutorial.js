// @flow

import React, { PureComponent } from 'react'
import {
  AsyncStorage,
  View,
  Image,
  StyleSheet,
  Animated,
} from 'react-native'
import { LinkButton } from '../../../common/buttons'
import { Text } from '../../../common/text'
import i18n from 'format-message'

type TutorialType = {
  id: string,
  text: string,
  image: any,
}

type Props = {
  tutorials: Array<TutorialType>,
}

type State = {
  hasLoaded: boolean,
  hasSeen: { [string]: boolean },
  currentTutorial: ?TutorialType,
}

const SPEED_GRADER_TUTORIAL_KEY = 'speed-grader-tutorial'

export default class Tutorial extends PureComponent {
  props: Props
  state: State
  opacity: Animated.Value

  constructor (props: Props) {
    super(props)

    this.state = {
      hasLoaded: false,
      hasSeen: {},
      currentTutorial: null,
    }

    this.opacity = new Animated.Value(0)
  }

  componentDidMount () {
    this.setup()
  }

  setup = async () => {
    let data = await AsyncStorage.getItem(SPEED_GRADER_TUTORIAL_KEY)
    let hasSeen = JSON.parse(data || '{}')

    this.setState({
      hasLoaded: true,
      hasSeen,
      currentTutorial: this.props.tutorials.find(({ id }) => !hasSeen[id]),
    })

    Animated.timing(
      this.opacity,
      {
        toValue: 1,
      }
    ).start()
  }

  onPress = async () => {
    let currentTutorial = this.state.currentTutorial
    if (!currentTutorial) return null

    let hasSeen = {
      ...this.state.hasSeen,
      [currentTutorial.id]: true,
    }
    AsyncStorage.setItem(SPEED_GRADER_TUTORIAL_KEY, JSON.stringify(hasSeen))
    let nextTutorial = this.props.tutorials.find(({ id }) => !hasSeen[id])

    await new Promise(resolve => {
      if (nextTutorial) return resolve()
      Animated.timing(
        this.opacity,
        {
          toValue: 0,
        }
      ).start(resolve)
    })

    this.setState({
      hasSeen,
      currentTutorial: nextTutorial,
    })
  }

  render () {
    let currentTutorial = this.state.currentTutorial
    if (!this.state.hasLoaded) return null
    if (!currentTutorial) return null

    return (
      <Animated.View
        style={[
          styles.tutorial,
          {
            opacity: this.opacity,
          },
        ]}
      >
        <View style={styles.box}>
          <Text style={styles.text}>{currentTutorial.text}</Text>
          <Image style={styles.image} source={currentTutorial.image} />
          <LinkButton
            style={styles.button}
            onPress={this.onPress}
            testID={`tutorial.button-${currentTutorial.id}`}
          >
            {i18n('OK')}
          </LinkButton>
        </View>
      </Animated.View>
    )
  }
}

const styles = StyleSheet.create({
  tutorial: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  box: {
    backgroundColor: 'white',
    borderRadius: 8,
    padding: 16,
    marginHorizontal: 40,
    alignItems: 'center',
    shadowColor: 'black',
    shadowRadius: 10,
    shadowOpacity: 0.5,
    shadowOffset: { width: 0, height: 10 },
  },
  text: {
    fontWeight: '600',
    fontSize: 18,
  },
  button: {
    fontSize: 16,
    fontWeight: '600',
  },
  image: {
    marginVertical: 24,
  },
})
