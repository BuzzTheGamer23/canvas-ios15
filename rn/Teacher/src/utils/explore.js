/* @flow */

function hasProp (object: any, key: string, value: any): boolean {
  if (object instanceof Object) {
    const props = object.props
    if (props && props instanceof Object) {
      for (var prop in props) {
        if (prop === key && props[prop] === value) {
          return true
        }
      }
    }
  }

  return false
}

function selectByProp (object: any, propKey: string, value: any): ?any {
  if (!(typeof object === 'object')) {
    return null
  }

  if (hasProp(object, propKey, value)) {
    return object
  }

  let result = null

  for (var key in object) {
    if (object[key] instanceof Array) {
      object[key].forEach((e) => {
        result = selectByProp(e, propKey, value)
        if (result) {
          return result
        }
      })
    }

    result = selectByProp(object[key], propKey, value)
    if (result) {
      return result
    }
  }

  return result
}

class ComponentExplorer {
  component: any

  constructor (component: any) {
    this.component = component
  }

  selectByProp (propKey: string, value: any): ?any {
    return selectByProp(this.component, propKey, value)
  }

  selectByID (id: string): ?any {
    return this.selectByProp('testID', id)
  }
}

export default function explore (component: any): ComponentExplorer {
  return new ComponentExplorer(component)
}
