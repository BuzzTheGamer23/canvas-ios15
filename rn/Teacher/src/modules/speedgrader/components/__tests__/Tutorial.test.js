// @flow

import React from 'react'
import { Animated } from 'react-native'
import renderer from 'react-test-renderer'
import Tutorial from '../Tutorial'
import Images from '../../../../images'
import explore from '../../../../../test/helpers/explore'

jest
  .mock('AsyncStorage', () => ({
    getItem: jest.fn(() => Promise.resolve()),
    setItem: jest.fn(() => Promise.resolve()),
  }))
  .mock('Animated', () => ({
    Value: jest.fn(),
    timing: () => ({ start: jest.fn() }),
    View: 'Animated.View',
  }))
  .mock('react-native-button', () => 'Button')

let defaultProps = {
  tutorials: [{
    id: '1',
    text: 'Some text',
    image: Images.speedGrader.swipe,
  }],
}

describe('Tutorial', () => {
  beforeEach(() => jest.resetAllMocks())
  it('will render null before data from async storage has been retrieved', () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    ).toJSON()

    expect(tree).toBeNull()
  })

  it('will render the tutorial after data from async storage has been retrieved', async () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )
    tree.getInstance().setState({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial: defaultProps.tutorials[0],
    })

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('will figure out what tutorial to show in setup', async () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )

    await tree.getInstance().setup()

    expect(tree.getInstance().state).toEqual({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial: defaultProps.tutorials[0],
    })
  })

  it('will update hasSeen when onPress is called', async () => {
    let oldTiming = Animated.timing
    let start = jest.fn()
    Animated.timing = () => ({ start })

    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )
    tree.getInstance().setState({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial: defaultProps.tutorials[0],
    })

    let button = explore(tree.toJSON()).selectByID('tutorial.button-1') || {}
    let promise = button.props.onPress()

    start.mock.calls[0][0]()
    await promise

    expect(tree.getInstance().state).toEqual({
      hasLoaded: true,
      hasSeen: { '1': true },
      currentTutorial: undefined,
    })

    Animated.timing = oldTiming
  })
})
