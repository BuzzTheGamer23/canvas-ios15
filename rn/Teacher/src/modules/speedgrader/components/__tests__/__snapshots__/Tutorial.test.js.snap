// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tutorial will render the tutorial after data from async storage has been retrieved 1`] = `
<Animated.View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.5)",
        "bottom": 0,
        "display": "flex",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "opacity": mockConstructor {},
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 8,
        "marginHorizontal": 40,
        "padding": 16,
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 10,
          "width": 0,
        },
        "shadowOpacity": 0.5,
        "shadowRadius": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontFamily": ".SFUIDisplay",
          },
          Object {
            "color": "#2D3B45",
            "fontSize": 16,
          },
          Object {
            "fontSize": 18,
            "fontWeight": "600",
          },
          Object {
            "fontFamily": ".SFUIDisplay-semibold",
          },
        ]
      }
    >
      Some text
    </Text>
    <Image
      source={1}
      style={
        Object {
          "marginVertical": 24,
        }
      }
    />
    <Button
      onPress={[Function]}
      style={
        Object {
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
      testID="tutorial.button-1"
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontFamily": ".SFUIDisplay",
            },
            Object {
              "color": "#2D3B45",
              "fontSize": 16,
            },
            Array [
              Object {
                "color": "#374A59",
                "fontSize": 14,
              },
              Object {
                "fontFamily": ".SFUIDisplay-bold",
              },
              Object {
                "color": "#374A59",
              },
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              },
            ],
            Object {
              "fontFamily": ".SFUIDisplay-semibold",
            },
          ]
        }
      >
        OK
      </Text>
    </Button>
  </View>
</Animated.View>
`;
